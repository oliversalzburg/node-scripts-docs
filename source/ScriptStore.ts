import fs from "node:fs/promises";
import path from "node:path";
import { isNil, mustExist } from "@oliversalzburg/js-utils/data/nil.js";
import { InvalidOperationError } from "@oliversalzburg/js-utils/errors/InvalidOperationError.js";

export const SCRIPTS_METADATA_DEFAULT_FILENAME = "package.json-scripts.json";

/**
 * A single entry in the script store.
 */
export class ScriptStoreEntry {
  /**
   * The path where the manifest is located.
   */
  manifestPath: string;

  /**
   * The name of the project this script exists in.
   */
  projectName: string;

  /**
   * The name of the script itself.
   */
  scriptName: string;

  /**
   * The code of the script.
   */
  scriptCode: string;

  /**
   * Is this a global script?
   * Scripts are usually global when their name includes a colon.
   */
  isGlobal: boolean;

  /**
   * Is this a script from the root manifest of the monorepo?
   */
  isRootManifest: boolean;

  /**
   * The user-authored description of the script.
   */
  description?: string | undefined;

  /**
   * Constructs a new script store entry.
   * @param manifestPath - The path where the manifest is located.
   * @param projectName - The name of the project this script exists in.
   * @param scriptName - The name of the script itself.
   * @param scriptCode - The code of the script.
   * @param isGlobal - Is this a global script?
   * @param isRootManifest - Is this a script from the root manifest of the monorepo?
   */
  constructor(
    manifestPath: string,
    projectName: string,
    scriptName: string,
    scriptCode: string,
    isGlobal: boolean,
    isRootManifest = false,
  ) {
    this.manifestPath = manifestPath;
    this.projectName = projectName;
    this.scriptName = scriptName;
    this.scriptCode = scriptCode;
    this.isGlobal = isGlobal;
    this.isRootManifest = isRootManifest;
  }
}

/**
 * The script store.
 */
export class ScriptStore {
  /**
   * Our default comment.
   */
  readonly comment = "This file is generated by node-scripts-docs.";

  /**
   * The actual scripts in the store.
   */
  scripts = new Array<ScriptStoreEntry>();

  /**
   * The current schema version of the script store.
   */
  readonly version: number = 1;

  #filename: string;
  #rootDirectory: string;

  /**
   * Retrieve all scripts that are global.
   * @returns All scripts that are global.
   */
  get globalScripts() {
    return this.scripts.filter(scriptMeta => scriptMeta.isGlobal);
  }

  /**
   * Constructs a new script store.
   * @param rootDirectory - The root directory of the project.
   * @param filename - The name of the file where we persist script metadata.
   */
  constructor(rootDirectory: string, filename = SCRIPTS_METADATA_DEFAULT_FILENAME) {
    this.#rootDirectory = rootDirectory;
    this.#filename = filename;
  }

  /**
   * Add a new script to the store.
   * @param manifestPath - The path where the manifest is located.
   * @param projectName - The name of the project this script exists in.
   * @param scriptName - The name of the script itself.
   * @param scriptCode - The code of the script.
   * @param isGlobal - Is this a global script?
   * @param isRootManifest - Is this a script from the root manifest of the monorepo?
   */
  add(
    manifestPath: string,
    projectName: string,
    scriptName: string,
    scriptCode: string,
    isGlobal: boolean,
    isRootManifest = false,
  ) {
    this.scripts.push(
      new ScriptStoreEntry(
        manifestPath,
        projectName,
        scriptName,
        scriptCode,
        isGlobal,
        isRootManifest,
      ),
    );

    // Assume consistent order is best preserved on insert.
    this.scripts.sort(
      (a, b) =>
        a.scriptName.localeCompare(b.scriptName) || a.projectName.localeCompare(b.projectName),
    );
  }

  /**
   * Save the store metadata to disc.
   * @param filename - The file to save the metadata to.
   * @returns A promise that is fulfilled once the metadata has been written to disc.
   */
  save(filename?: string): Promise<void> {
    return fs.writeFile(
      path.resolve(this.#rootDirectory, filename ?? this.#filename),
      JSON.stringify(this, undefined, 2),
    );
  }

  /**
   * Load script metadata from the given file.
   * @param filename - The file to load the metadata from.
   */
  async load(filename?: string) {
    const storeContent = await fs.readFile(
      path.resolve(this.#rootDirectory, filename ?? this.#filename),
      "utf-8",
    );
    const store = JSON.parse(storeContent) as ScriptStore;
    if (store.version !== this.version) {
      throw new Error("Unexpected version mismatch!");
    }
    this.scripts = mustExist(store.scripts);
  }

  /**
   * Generates a locator name for a script.
   * @param projectName - The name if the project.
   * @param scriptName - The name of the script.
   * @returns A locator name.
   */
  static makeScriptLocator(projectName: string, scriptName: string) {
    return `${scriptName}→${projectName}`;
  }

  /**
   * Parse a script locator into the project and script name parts.
   * @param locator - The script locator.
   * @returns The project and script name the locator refers to.
   */
  static parseScriptLocator(locator: string) {
    const parts = /([^→]+)→(.+)/.exec(locator);
    if (isNil(parts)) {
      throw new InvalidOperationError(`The locator '${locator}' is invalid.`);
    }

    return {
      projectName: parts[2],
      scriptName: parts[1],
    };
  }

  /**
   * Check if a store exists at the given location.
   * @param storePath - A path where a store might exist.
   * @returns `true` if the store exists; `false` otherwise.
   */
  static async exists(storePath: string) {
    try {
      await fs.stat(storePath);
    } catch {
      return false;
    }

    return true;
  }
}
