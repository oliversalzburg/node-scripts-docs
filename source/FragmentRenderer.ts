import { ScriptStoreEntry } from "./ScriptStore.js";

export const DOCUMENTATION_PENDING_DEFAULT = "_documentation pending_";

/**
 * Determine if the current description of a script is still at its default value.
 * @param description - The current description of the script.
 * @returns `true` if it's the default description; `false` otherwise.
 */
export const isDefaultDescription = (description: string) => {
  return description === DOCUMENTATION_PENDING_DEFAULT;
};

/**
 * Render a documentation fragment.
 * @param scriptMeta - The script metadata.
 * @param rawDescription - The description of the script.
 * @returns A documentation fragment for the provided script.
 */
export const makeDocumentation = (
  scriptMeta: ScriptStoreEntry,
  rawDescription?: string | undefined,
) => {
  const description = rawDescription ?? DOCUMENTATION_PENDING_DEFAULT;
  const descriptionIndented = description.replace(/^(?!\s*$)/gm, " ".repeat(4));

  const scriptDocs = `## ${scriptMeta.scriptName}

-   Project: \`${scriptMeta.projectName}\`
-   Source:

    \`\`\`shell
    ${scriptMeta.scriptCode}
    \`\`\`

-   Description:

${descriptionIndented}
`;

  return scriptDocs;
};
