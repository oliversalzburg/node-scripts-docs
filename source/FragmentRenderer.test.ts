import { isDefaultDescription, makeDocumentation } from "./FragmentRenderer.js";
import { ScriptStoreEntry } from "./ScriptStore.js";

it("identifies pending description", () => {
  expect(isDefaultDescription("_documentation pending_")).toBe(true);
  expect(isDefaultDescription("Some description")).toBe(false);
});

it("renders fragment with existing description", () => {
  const meta: ScriptStoreEntry = {
    description:
      "Build the latest sources and then use the build output to execute your command. Parameters are passed through by `npm exec`.",
    manifestPath: "package.json",
    projectName: "node-scripts-docs",
    scriptName: "docs:scripts",
    scriptCode: "npm exec -- nsd-debug --include-locals",
    isGlobal: true,
    isRootManifest: true,
  };
  const fragment = makeDocumentation(meta, meta.description);
  expect(fragment).toMatchSnapshot();
});

it("renders fragment without description", () => {
  const meta: ScriptStoreEntry = {
    manifestPath: "package.json",
    projectName: "node-scripts-docs",
    scriptName: "docs:scripts",
    scriptCode: "npm exec -- nsd-debug --include-locals",
    isGlobal: true,
    isRootManifest: true,
  };
  const fragment = makeDocumentation(meta, meta.description);
  expect(fragment).toMatchSnapshot();
});
