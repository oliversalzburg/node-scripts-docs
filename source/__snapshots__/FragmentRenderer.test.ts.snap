// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders fragment with existing description 1`] = `
"## docs:scripts

-   Project: \`node-scripts-docs\`
-   Source:

    \`\`\`shell
    npm exec -- nsd-debug --include-locals
    \`\`\`

-   Description:

    Build the latest sources and then use the build output to execute your command. Parameters are passed through by \`npm exec\`.
"
`;

exports[`renders fragment without description 1`] = `
"## docs:scripts

-   Project: \`node-scripts-docs\`
-   Source:

    \`\`\`shell
    npm exec -- nsd-debug --include-locals
    \`\`\`

-   Description:

    _documentation pending_
"
`;
